octavia_f5/controller/worker/tasks/f5_tasks.py

```diff
@@ -547,34 +547,67 @@ def revert(self, bigip: bigip_restclient.BigIPRestClient,
 
 class RemoveRouteDomain(task.Task):
     def execute(self, network: f5_network_models.Network,
-                bigip: bigip_restclient.BigIPRestClient):
-
-        paths = [
-            f"/mgmt/tm/net/route-domain/vlan-{network.vlan_id}",
-            f"/mgmt/tm/net/route-domain/net-{network.id}"
-        ]
+                bigip: bigip_restclient.BigIPRestClient,
+                existing_route_domain):
 
         """ Task to delete Route Domain """
-        res = None
-        for path in paths:
-            if bigip.get(path=path).ok:
-                res = bigip.delete(path=path)
-                break
-
-        if res and not res.ok:
-            LOG.warning("%s: Failed removing route domain for network_id=%s vlan_id=%s: %s",
-                        bigip.hostname, network.id, network.vlan_id, res.content)
+        if existing_route_domain is None:
+            return
+        res = bigip.delete(path=f"/mgmt/tm/net/route-domain/{existing_route_domain['fullPath']}")
+        res.raise_for_status()
 
+    @decorators.RaisesIControlRestError()
+    def revert(self,
+                bigip: bigip_restclient.BigIPRestClient,
+                existing_route_domain, result, *args, **kwargs):
+        if isinstance(result, failure.Failure):
+            # If this task failed it means that object was not removed
+            return
+        # Restore RouteDomain if it existed before
+        if existing_route_domain is not None:
+            res = bigip.post(
+                path='/mgmt/tm/net/route-domain',
+                json={
+                    'name': existing_route_domain['name'],
+                    'vlans': existing_route_domain['vlans'],
+                    'id': existing_route_domain['id']
+                }
+            )
+            res.raise_for_status()
 
 class RemoveVLAN(task.Task):
+
+    @decorators.RaisesIControlRestError()
     def execute(self, network: f5_network_models.Network,
-                bigip: bigip_restclient.BigIPRestClient):
+                bigip: bigip_restclient.BigIPRestClient,
+                existing_vlan: dict):
         """ Task to delete VLAN """
-        name = f'vlan-{network.vlan_id}'
-        res = bigip.delete(path=f"/mgmt/tm/net/vlan/~Common~{name}")
-        if not res.ok:
-            LOG.warning("%s: Failed removing VLAN for vlan_id=%s: %s",
-                        bigip.hostname, network.vlan_id, res.content)
+        if existing_vlan is None:
+            return
+        res = bigip.delete(path=f"/mgmt/tm/net/vlan/~Common~vlan-{network.vlan_id}")
+        res.raise_for_status()
+
+    @decorators.RaisesIControlRestError()
+    def revert(self,
+                bigip: bigip_restclient.BigIPRestClient,
+                existing_vlan: dict, result, *args, **kwargs):
+        if isinstance(result, failure.Failure):
+            # If this task failed it means that object was not removed
```

Comment:
I'm not entirely sure whether this is always true. So I would like to at least write a log line here stating that the worker is assuming the VLAN to still exist.

Reply:
It's working. Documentation: https://docs.openstack.org/taskflow/latest/user/arguments_and_results.html#revert-arguments 

```
If this task failed (ie do_something raised an exception) it will print "This task failed, exception:" and a exception message on revert. If this task finished successfully, it will print "do_something returned" and a representation of the do_something result.
``` 

And my tests are covering this situation.

Reply:
Log message was added